runtime/src/asset.rs

```diff
@@ -1774,4 +1783,30 @@ impl<T: Trait> Module<T> {
 
         Ok(())
     }
+
+    /// Adds a new document
+    fn add_document(ticker: Ticker, doc: Document) {
+        let new_nonce = Self::document_nonce() + 1u64;
+        <DocumentNonce>::put(&new_nonce);
+
+        let last_id = Self::last_document_id(&ticker);
+
+        if last_id > 0 {
+            // 0 means no previous link. 0 is the default value.
+            // Changing the last link to point to new link as next link.
+            <Documents<T>>::mutate((ticker.clone(), last_id), |last_doc| {
+                last_doc.next_link = new_nonce
+            });
+        }
+
+        let link = Link {
+            link_data: LinkData::Document(doc.name, doc.uri, doc.hash),
+            expiry: None,
+            next_link: 0,
+            previous_link: last_id,
+        };
+
+        <LastDocumentId>::insert(ticker.clone(), new_nonce);
+        <Documents<T>>::insert((ticker.clone(), new_nonce), link);
```

Comment:
I don't think there is a need to clone here as well.  and It is good to emit an event after adding and removing a document

Reply:
Asset module now leverages link infrastructure as suggested by @maxsam4 above. Links trigger  events with `link_id` which serves as doc_id.